File: src/lib/formats/mdl/components/texture-anim.ts
import { QuaternionRotation, Vector3 } from '@/lib/math/common';

import { Animation, animationToString } from './animation';

export interface TextureAnim {
  id: number
  translation?: Animation<Vector3>;
  scaling?: Animation<Vector3>;
  rotation?: Animation<QuaternionRotation>;
}

export function textureAnimsToString(textureAnims: TextureAnim[]): string {
  if (textureAnims.length === 0) return '';
  return `TextureAnims ${textureAnims.length} {
    ${textureAnims.map((texAnim) => `TVertexAnim {
      ${animationToString('Translation', texAnim.translation)}
      ${animationToString('Rotation', texAnim.rotation)}
      ${animationToString('Scaling', texAnim.scaling)}
    }`).join('\n')}
  }`;
}

File: golang/internal/formats/mdl/components/texture_anim.go
package components

import "strings"

type TextureAnim struct {
	ID          int
	Translation *Animation
	Scaling     *Animation
	Rotation    *Animation
}

func TextureAnimsToString(textureAnims []TextureAnim) string {
	if len(textureAnims) == 0 {
		return ""
	}
	var b strings.Builder
	b.WriteString("TextureAnims ")
	b.WriteString(FVal(float64(len(textureAnims))))
	b.WriteString(" {\n")
	for _, texAnim := range textureAnims {
		b.WriteString("TVertexAnim {\n")
		b.WriteString(AnimationToString("Translation", texAnim.Translation))
		b.WriteString(AnimationToString("Rotation", texAnim.Rotation))
		b.WriteString(AnimationToString("Scaling", texAnim.Scaling))
		b.WriteString("}\n")
	}
	b.WriteString("}")
	return b.String()
}
